core/fifo_cache.go

```diff
@@ -0,0 +1,39 @@
+// (c) 2021, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package core
+
+import "github.com/ava-labs/avalanchego/utils/linkedhashmap"
+
+// FIFOCache evicts the oldest element added to it after [limit] items are
+// added.
+//
+// TODO: Move this to AvalancheGo
+type FIFOCache[K, V any] struct {
+	limit int
+
+	m linkedhashmap.LinkedHashmap[K, V]
+
+	// [Linkedhashmap] is thread-safe, so no additional locking is required
+}
+
+func NewFIFOCache[K comparable, V any](limit int) *FIFOCache[K, V] {
+	return &FIFOCache[K, V]{
+		limit: limit,
+		m:     linkedhashmap.New[K, V](),
+	}
+}
+
+func (f *FIFOCache[K, V]) Put(key K, val V) {
+	if f.m.Len() >= f.limit {
```

Comment:
This should be better protected before merging. It isn't a race because we only call `Put` in a single thread.